### PR TITLE
Time dependent flow boundary

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/439.

### CODE REVIEW

File: core/src/create.jl
Comment (Hofer-Julian) [inline core/src/create.jl:9]: I think ∉ was more readable to be honest.
Do you agree @visr?

File: core/src/create.jl
Comment (Hofer-Julian) [inline core/src/create.jl:220]:
```suggestion
    msg = "FlowBoundary cannot be in both static and time tables"
```

File: core/src/create.jl
Comment (Hofer-Julian) [inline core/src/create.jl:223]: Why do you need to collect here?

File: core/src/create.jl
Comment (Hofer-Julian) [inline core/src/create.jl:229]: Where is this variable used?

File: core/src/create.jl
Comment (Hofer-Julian) [inline core/src/create.jl:231]: Nice, didn't know this syntax before.

File: core/src/solve.jl
Comment (Hofer-Julian) [inline core/src/solve.jl:233]: Document `time` here

File: docs/core/usage.qmd
Comment (Hofer-Julian) [inline docs/core/usage.qmd:329]:
```suggestion
The table must by sorted by time, and per time it must be sorted by `node_id`.
```

File: python/ribasim/ribasim/node_types/flow_boundary.py
Comment (Hofer-Julian) [inline python/ribasim/ribasim/node_types/flow_boundary.py:25]: Was that the solution for the empty dataframe problem we encountered yesterday?

**Review by Hofer-Julian — COMMENTED** [review]
Looks good to me. Maybe @visr can have another look as well.

File: core/src/create.jl
Comment (SouthEndMusic) [reply core/src/create.jl:223]: Because I need to convert from `Set` to `Vector` for other operations later to work, I think `findfirst(static_node_ids .== node_id)`. It's a bit silly though, I'll use `static.node_id` and `time.node_id` directly.

File: core/src/create.jl
Comment (SouthEndMusic) [reply core/src/create.jl:229]: Nowehere :^)

Comment (Hofer-Julian) [reply core/src/create.jl:229]: We really need some kind of static analyzer for that

File: python/ribasim/ribasim/node_types/flow_boundary.py
Comment (SouthEndMusic) [reply python/ribasim/ribasim/node_types/flow_boundary.py:25]: No it wasn't, I wanted an empty `FlowBoundary / static` but I added a dummy node in there. In the end that was fine since I needed to test having both static and dynamic nodes. The problem persists though, I will post about it in the dev channel.

File: core/src/create.jl
Comment (visr) [inline core/src/create.jl:216]: Looks like static and time tables are loaded twice?

File: core/src/create.jl
Comment (visr) [inline core/src/create.jl:232]: This line is correct but can be improved in two ways. `a .== b` creates an intermediate boolean vector that we can avoid, e.g. by `findfirst(==(node_id), static.node_id)`. `==(node_id)` creates a function that takes one argument that checks if it is equal to `node_id`, similar to `x -> x == node_id`.

But we also know from `load_structvector` that `static.node_id` is sorted, so we can use binary search to speed things up, which will help if there are many FlowBoundaries.

```suggestion
            static_idx = searchsortedfirst(static.node_id, node_id)
```

File: docs/core/usage.qmd
Comment (visr) [inline docs/core/usage.qmd:330]:
```suggestion
With this the flow rates can be updated over time.
```

File: python/ribasim/ribasim/models.py
Comment (visr) [inline python/ribasim/ribasim/models.py:171]: Nit, but it's easier to read if this is next to the FlowBoundaryStatic. Perhaps we should sort all these lines alphabetically (your editor can do that).

File: core/test/time.jl
Comment (visr) [inline core/test/time.jl:10]: It's fine either way, but FYI if you get tired of the dots, you can use the `@.` macro: `@. y = x + 3 * sin(x)` does everything elementwise.

**Review by visr — CHANGES_REQUESTED** [review]
Some small improvements, doc fixes, but otherwise this looks good to me!

File: core/src/create.jl
Comment (SouthEndMusic) [reply core/src/create.jl:232]: Neat, I found some more instances of this

File: core/test/time.jl
Comment (SouthEndMusic) [reply core/test/time.jl:10]: It seems that your usage of the macro is not quite right, it should be `y = @. x + 3 * sin(x)`.

**Comment by SouthEndMusic** [comment]
@visr how do I properly handle 'Changes requested'?

**Comment by Hofer-Julian** [comment]
> @visr how do I properly handle 'Changes requested'?

You wait until @visr gives an approval. Let's see if I can do that too.

**Review by Hofer-Julian — APPROVED** [review]

**Comment by Hofer-Julian** [comment]
No, seems like Martijn has to approve himself.

File: core/test/time.jl
Comment (visr) [reply core/test/time.jl]: `@. y = x + 3 * sin(x)` also works but only if you already have a vector y, then it fills it in place, if you don't, then indeed your version is the right one to use.

**Review by visr — APPROVED** [review]

**Pull request merged**
